### PR TITLE
tool: find curlrc in Unicode home on Windows

### PR DESCRIPTION
**Pull request by hunterinvariants**
Read the home-directory environment variables with the Windows Unicode API when the Unicode curl tool searches for its default config file. Convert only those directory values to UTF-8 before using curl's existing UTF-8-aware file functions.

This keeps curl_getenv and its encoding behavior unchanged for libcurl and third-party library callers.

Add a tool unit test that creates `_curlrc` under a Unicode CURL_HOME path. The test fails before this change and passes afterward in a native Windows 11 Unicode build.

Verified with 32-bit and 64-bit MinGW builds, a MinGW Unicode debug build, checksrc, and the native Windows 11 regression test.

### CODE REVIEW

**Comment by jay** [comment]
To me it seems unlikely that the user's home directory will not be available in the local encoding, have you seen this happen

**Comment by hunterinvariants** [comment]
I reproduced it on Windows 11 using `CURL_HOME=C:\temp\curl-home-漢字-ä`. The existing Unicode build did not load `_curlrc` from that directory, while the same test with an ASCII-only `CURL_HOME` did. With this change, the Unicode path is found and the config is loaded.

I have not reproduced it with an actual `USERPROFILE` created by Windows. The concrete case I observed is an explicitly configured `CURL_HOME` outside the active code page.

**Comment by testclutch** [comment]
Analysis of PR #22549 at [50f4a3cd](https://github.com/curl/curl/commit/50f4a3cd362ccbcd431a63e2f318ca4e06862781):

[Test 3017 failed](https://github.com/curl/curl/actions/runs/31518781157/job/93870412242), but it has been 53.9% flaky lately, so it's *probably NOT* a fault of the PR.  Note that this CI job has had a number of other flaky tests recently (2, to be specific) so it may be that this failure is rather a systemic issue with this job and not with this specific PR.

[Test 2300 failed](https://github.com/curl/curl/actions/runs/31518781152/job/93870458118), but it has been 10.0% flaky lately, so it's *probably NOT* a fault of the PR.

###### Generated by [Testclutch](https://testclutch.curl.se/)

File: src/tool_findfile.c [hunk 1]
```diff
@@ -60,6 +60,48 @@ static const struct finder conf_list[] = {
...
+    if(rc < size) {
+      value = curlx_convert_wchar_to_UTF8(buf);
+      break;
+    }
+    size = rc;
+  }
+
+  curlx_free(buf);
+  curlx_free(name);
+  return value;
```
Comment (vszakats) [inline src/tool_findfile.c:103]: Not a fan of copy-pasting the libcurl implementation with Unicode added on, into src.
This makes the codebase ending up with 3 different getenv implementations, all with
subtly different behavior, and weird dependencies depending on build.

Personally I think that solving Unicode issues by cherry-picking individual cases like
this may end up causing more confusion (and future issues) than they may solve.

I wonder if there a better solution to this.

**Comment by vszakats** [comment]
> I reproduced it on Windows 11 using `CURL_HOME=C:\temp\curl-home-漢字-ä`. The existing Unicode build did not load `_curlrc` from that directory, while the same test with an ASCII-only `CURL_HOME` did. With this change, the Unicode path is found and the config is loaded.
> 
> I have not reproduced it with an actual `USERPROFILE` created by Windows. The concrete case I observed is an explicitly configured `CURL_HOME` outside the active code page.

Why would someone want do override the system `USERPROFILE`, or set a `CURL_HOME` with the wrong encoding in a real-world environment?

**Comment by hunterinvariants** [comment]
`CURL_HOME` is the documented override for curl's config directory, so applications, portable installations and automated environments may intentionally place it outside `USERPROFILE`.

The value is not deliberately set using a “wrong encoding”. Windows environment variables are Unicode; the loss occurs when a Unicode build reads the value through the ANSI API and the path contains characters outside the active code page.

That said, my reproduced case uses an explicit `CURL_HOME`, not a naturally created `USERPROFILE`. If supporting such override paths is not considered worthwhile, I am fine with closing this PR.

File: src/tool_findfile.c (hunk 1 above)
Comment (hunterinvariants) [reply src/tool_findfile.c:103]: Alright, duplicating the environment reader in the tool is not a good direction. I am looking at moving the shared allocation and retry logic into one internal helper with an explicit local-codepage or UTF-8 result, while keeping the existing curl_getenv behavior unchanged.

Comment (hunterinvariants) [reply src/tool_findfile.c:103]: A possible shared solution would be an internal `curlx` environment reader with an explicit result encoding. The existing public `curl_getenv()` wrapper would request the current Windows code page, preserving its behavior, while `tool_findfile.c` would request UTF-8 for paths in Unicode builds. The allocation, retry, empty-value and size-limit handling would then have one implementation shared by libcurl and the tool. Would that direction address your concern, or would you prefer to leave the current behavior unchanged?

Comment (vszakats) [reply src/tool_findfile.c:103]: Creating a `curlx_getenv()` in curlx and calling that from other places,
including `curl_getenv()` would be a good direction, probably, yes.
Reminder that `curl_getenv()` is a public API, and needs to remain
compatible. This step would perhaps better done as a separate,
perparation step? If there is an agreement on it.

My concern remains that libcurl/curl internally has no concept of knowing
which string is in what encoding. Such curlx API with an encoding option
would add such concept, but I miss the planning around this, and
encoding/codepages can go messy quickly without it. It's also somewhat
of a annoyance on all non-Windows platforms, and any solution would
need to blend in there smoothly in an ideal case.

Comment (hunterinvariants) [reply src/tool_findfile.c:103]: Agreed. I would keep the abstraction out of this PR and treat it as a separate preparation change.

A possible contract would be:

- `CURLX_ENV_NATIVE` preserves the current `curl_getenv()` behavior. On Windows this means the active code page; on other platforms it returns the environment bytes unchanged.
- `CURLX_ENV_UTF8` uses the wide Windows environment API and converts the result to UTF-8. On non-Windows platforms it returns the environment bytes unchanged, since the OS does not provide encoding metadata.
- Missing and empty variables, allocation ownership and failure behavior remain compatible with `curl_getenv()`.
- The variable name is supplied as UTF-8 for the UTF-8 mode and retains the existing interpretation in native mode.

The preparation PR would add and test the internal helper, then make the public `curl_getenv()` call it in native mode without changing observable behavior. A later tool-only PR could use UTF-8 mode for config path discovery in Windows Unicode builds.

Would this be a suitable contract, including the byte-preserving behavior on non-Windows, or would you prefer the UTF-8 mode to be Windows-specific?

Comment (vszakats) [reply src/tool_findfile.c:103]: For this particular case, probably. Though maybe trying to avoid adding
two constants (effectively a boolean) could simplify?

What I'm still missing is the big picture. (How could anyone follow that
the string read like this is in a different encoding than all others? How
do we follow where it's ending up, if it's used a filename, passed to an
API, used in I/O, how to ensure this is honored in future updates. May
it break existing installations, `CURL_HOME` values?)

E.g. `findfile()` is used to load `knownhosts` too. This string is passed to
libcurl, which then passes it to libssh or libssh2. libssh2 just received
support for Unicode (not in a release yet), with some of the same issues
as in libcurl. How do we know this UTF-8 string finds its way into these 
libraries, and they end up consuming them as such and passing it to the
OS as intended?

Also noting that finding the wrong `knownhosts` and `.curlrc` may
upgrade to a security problem, if these intrepretations are mixed up.

I would personally not start on this road without a plan.

It may be an unpopular stance, but till these clear up, I think it's just
easier not to put `.curlrc` (and `knownhosts`) on a Unicode path. If it
doesn't break curl with this fix in (assuming it's a Unicode build, with its
other quirks), it most certainly breaks something else. Is this ideal? No.
But Windows' encoding/codepage/unicode "story" made it practically
impossible, or at least very costly/cumbersome to support this in
a portable codebase, which handles strings as transparent/raw bytes
and interprets them as UTF-8 where necessary (FS, I/O).

Comment (hunterinvariants) [reply src/tool_findfile.c:103]: There are only two `findfile()` callers. I can leave the existing function unchanged for knownhosts and add a config-specific path for `.curlrc`. Only that path would read the home directory as Unicode and return UTF-8 for `curlx_fopen()`. This would not change `curl_getenv()`, `CURLOPT_SSH_KNOWNHOSTS`, or anything passed to libssh/libssh2. Would that narrower approach make sense?

Comment (vszakats) [reply src/tool_findfile.c:103]: Wouldn't this mean that `CURL_HOME` and sibling envs would be interpreted as different encodings/strings depending on context?

**Comment by jay** [comment]
IMO we should skip this. It is unlikely to happen and, though it could happen, there is a maintenance cost that comes with all additional code we add. Also the idea discussed in one of the review comments to support changing different versions of curl getenv via environment variables seems complicated.

I think Unicode compatibility is best increased by changing the Unicode build of the app's local codepage to UTF-8 as we have discussed elsewhere. By app I am referring to the curl tool and not libcurl. That would make curl_getenv return UTF-8 encoded and solve the issues like the home directory is not in the local encoding (because the local encoding as seen by the app would be UTF-8 encoded Unicode so the directory name would always be available).

**Comment by vszakats** [comment]
> IMO we should skip this. It is unlikely to happen and, though it could happen, there is a maintenance cost that comes with all additional code we add. Also the idea discussed in one of the review comments to support changing different versions of curl getenv via environment variables seems complicated.

I agree to skip this PR.

Perhaps also migrate/merge the KNOWN_BUGS item(s) into a general Unicode bug.
Handling them string by string doesn't seem to work to me.

File: src/config2setopts.c [hunk 2]
```diff
@@ -229,7 +229,8 @@ static CURLcode ssh_setopts(struct OperationConfig *config, CURL *curl,
   if(!config->insecure_ok) {
     char *known = config->knownhosts;
     if(!known) {
-      char *found = findfile(".ssh/known_hosts", FALSE);
+      /* the path goes to libssh2/libssh, which expect the current locale */
```
Comment (vszakats) [inline src/config2setopts.c:232]: What libssh2/libssh expect is undefined. They may or may not expect the
current locale, and may or may not understand UTF-8. libcurl doesn't
know, and probably the dependencies don't document it (libssh2 certainly
does not). Even in the case they do expect a specific encoding, it's likely
conditional (depending on build settings, runtime settings, specific API,
version).

It's not changing the conclusion, just highlights one of the challenges.

**Comment by hunterinvariants** [comment]
Thanks, agreed, and I am closing this. Setting the app's active code page to UTF-8 fixes the same case without adding a second encoding path to maintain, which is the better trade.

I would like to try that route, and one detail seems worth settling first. `curl.exe` currently ships no manifest: 9b0467b169 removed it, and that removal was load-bearing at the time — dropping our own manifest let the toolchain's auto-generated SxS manifest through again, which is what fixed VS2008 shared builds. VS2008 is no longer in CI, but the shape of the constraint stands: an `activeCodePage` manifest would have to coexist with the toolchain-generated one rather than replace it, across CMake, autotools and winbuild.

Two further things I would want to get right, not objections: the setting applies from Windows 10 1903 onwards, so older targets keep the current behaviour; and while you mean the tool and not libcurl, a manifest applies to the process, so libcurl linked into `curl.exe` would see UTF-8 as its local encoding too. I assume that is intended, since it is what makes the home directory resolve.

Is there an existing issue or discussion I should attach this to? If the approach sounds right, I am happy to work it up. The unit test here (`findfile()` with a non-ASCII home directory) would serve unchanged as a regression test for it.

**Comment by vszakats** [comment]
> Thanks, agreed, and I am closing this. Setting the app's active code page to UTF-8 fixes the same case without adding a second encoding path to maintain, which is the better trade.
> 
> I would like to try that route, and one detail seems worth settling first. `curl.exe` currently ships no manifest: [9b0467b](https://github.com/curl/curl/commit/9b0467b169e08b70077f09313975d4eac14a9930) removed it, and that removal was load-bearing at the time — dropping our own manifest let the toolchain's auto-generated SxS manifest through again, which is what fixed VS2008 shared builds. VS2008 is no longer in CI, but the shape of the constraint stands: an `activeCodePage` manifest would have to coexist with the toolchain-generated one rather than replace it, across CMake, autotools and winbuild.
> 
> Two further things I would want to get right, not objections: the setting applies from Windows 10 1903 onwards, so older targets keep the current behaviour; and while you mean the tool and not libcurl, a manifest applies to the process, so libcurl linked into `curl.exe` would see UTF-8 as its local encoding too. I assume that is intended, since it is what makes the home directory resolve.

I don't think we can ignore libcurl, and fix curl-tool-only. Apps use
libcurl, and those apps may also need Unicode support, with an ABI
and semantics to rely on. curl being one such libcurl user.

FTR, I also don't agree with the W + UTF-8 approach. MS designed
this for ANSI + UTF-8, and avoiding translations would be an
important objective IMO. Though it'd be also useful to understand
the pros and cons of either approaches, and deciding on which one
to support, or support both and let everyone experiment.

> Is there an existing issue or discussion I should attach this to? If the approach sounds right, I am happy to work it up. The unit test here (`findfile()` with a non-ASCII home directory) would serve unchanged as a regression test for it.

Past discussions/Issues/PRs with the Unicode label (there aren't that many):
https://github.com/curl/curl/pulls?q=label%3AUnicode+is%3Aclosed
https://github.com/curl/curl/issues?q=is%3Aissue%20label%3AUnicode
https://github.com/curl/curl/discussions?discussions_q=label%3AUnicode

The most recent one: https://github.com/curl/curl/discussions/22348

This is an earlier attempt (of mine) to enable A + UTF-8, from 2023:
https://github.com/curl/curl/pull/11482

As for the manifest, indeed, the VS2008 thing is no longer an issue, but
setting a manifest without disturbing other tools wanting to set their own
seems to remain one. Also not to break portability and cross-buildability.

**Comment by jay** [comment]
To be clear my position is UTF-8 encoding for only Unicode builds of curl (ie UNICODE && _UNICODE defined), not all builds, as most recently described in [this post](https://github.com/curl/curl/discussions/22348#discussioncomment-17684149) in 22348. This would involve adding the active code page manifest for just those builds, I assume by executing a different mt command to get the different manifest or from an rc file like this:

~~~c
#ifdef UNICODE
1 RT_MANIFEST "Manifest-with-activecodepage.manifest"
#else
1 RT_MANIFEST "Manifest-without-activecodepage.manifest"
#endif
~~~

And as discussed in the post UTF-8 would need to be set for the CRT as well which would be `setlocale(LC_ALL, "");` gated on _UNICODE. The post goes into detail about this, there is some code and demonstration how this works in CRT.

Those two things would improve the curl Unicode situation.

> FTR, I also don't agree with the W + UTF-8 approach. MS designed
> this for ANSI + UTF-8, and avoiding translations would be an
> important objective IMO. Though it'd be also useful to understand
> the pros and cons of either approaches, and deciding on which one
> to support, or support both and let everyone experiment.

Windows already does A<->W translations a lot internally and there's really no avoiding that unless we redesign for wide characters. With the code page set it's all Unicode whether it's UTF-16 (W) or UTF-8 (A). I'm not worried about the extra translations, as it's overhead Windows already does anyway. Also we need the existing code at least for legacy as I've already mentioned.

**Comment by hunterinvariants** [comment]
@jay Thanks, that clarifies the intended scope. If there is enough agreement on the manifest plus `setlocale()` approach, I would be interested in implementing it.

As I understand it, Unicode builds would get the `activeCodePage` manifest and call `setlocale(LC_ALL, "")`, both gated on `_UNICODE`/`UNICODE`. Non-Unicode builds would keep their current behavior. The build integration would need to work with CMake, autotools and winbuild without regressing the generated manifest handling.

Is that the direction you both want, or is there anything still to settle before I start?

**Comment by jay** [comment]
We could never agree on a lot of these changes, you can see that in the discussions we have had about it, which is why I never proposed this myself in a PR. You're welcome to propose it but I can't guarantee it will make it. To me the broader Windows Unicode issue is like a weight-of-the-world type issue like how much can we really do when old MS design decision led to this. To be fair though they've come a long way and now support UTF-8. At the very least it should be an improvement, even if we can't agree on what is most correct, scope, etc.

**Comment by vszakats** [comment]
> To be clear my position is UTF-8 encoding for only Unicode builds of curl (ie UNICODE && _UNICODE defined), not all builds, as most recently described in [this post](https://github.com/curl/curl/discussions/22348#discussioncomment-17684149) in 22348. This would involve adding the active code page manifest for just those builds, I assume by executing a different mt command to get the different manifest or from an rc file like this:
> 
> ```c
> #ifdef UNICODE
> 1 RT_MANIFEST "Manifest-with-activecodepage.manifest"
> #else
> 1 RT_MANIFEST "Manifest-without-activecodepage.manifest"
> #endif
> ```

The `UNICODE` macro should not be the trigger to enable this
manifest. It should be a separate setting. They may be enabled
together, but should not be force-bound together IMO.

> Windows already does A<->W translations a lot internally and there's really no avoiding that unless we redesign for 
> wide characters. With the code page set it's all Unicode whether it's UTF-16 (W) or UTF-8 (A). I'm not worried about 
> the extra translations, as it's overhead Windows already does anyway. Also we need the existing code at least for 
> legacy as I've already mentioned.

Until these assumptions are tried and tested we don't know.
A single extra translation can break things, so they do count,
Windows code we can't change, curl code we can.
Also, why do you want to block this combination at any opportunity
that arises?

**Comment by jay** [comment]
> The `UNICODE` macro should not be the trigger to enable this manifest. It should be a separate setting. They may be enabled together, but should not be force-bound together IMO.

The way I think about this is that setting the active code page to UTF-8 means setting the active code page to Unicode. curl Unicode builds are ifdef UNICODE so it would make sense to check that macro.

> Until these assumptions are tried and tested we don't know.
> A single extra translation can break things, so they do count,
> Windows code we can't change, curl code we can.

UTF-8 (A) and UTF-16 (W) are both encodings for Unicode and lossless converting between the two. More use of the A API to stop conversions is just not how Windows works internally. Windows A functions wrap W functions because internally Windows uses UTF-16. Whether we convert it or Windows converts it. That was a design decision they made long ago. Though I don't have the source code to Windows to prove this.

> Also, why do you want to block this combination at any opportunity
> that arises?

Legacy as discussed in my many responses in these issues. I thought we had a good discussion in #22348 about this. I don't think it would be a good idea to restate more of it here ad nauseam.
